tests/subsys/audio_module/src/functional_test.c

```diff
@@ -0,0 +1,1065 @@
+			-EALREADY);
+	test_start_stop(false, &ft_null, AUDIO_MODULE_STATE_RUNNING, AUDIO_MODULE_STATE_STOPPED, 0);
+	test_start_stop(false, &ft_null, AUDIO_MODULE_STATE_CONFIGURED,
+			AUDIO_MODULE_STATE_CONFIGURED, -EALREADY);
+}
+
+ZTEST(suite_audio_module_functional, test_stop_fnct)
+{
+	test_start_stop(false, &ft_pop, AUDIO_MODULE_STATE_STOPPED, AUDIO_MODULE_STATE_STOPPED,
+			-EALREADY);
+	test_start_stop(false, &ft_pop, AUDIO_MODULE_STATE_RUNNING, AUDIO_MODULE_STATE_STOPPED, 0);
+	test_start_stop(false, &ft_pop, AUDIO_MODULE_STATE_CONFIGURED,
+			AUDIO_MODULE_STATE_CONFIGURED, -EALREADY);
+}
+
+ZTEST(suite_audio_module_functional, test_start_fnct_null)
+{
+	test_start_stop(true, &ft_null, AUDIO_MODULE_STATE_RUNNING, AUDIO_MODULE_STATE_RUNNING,
+			-EALREADY);
+	test_start_stop(true, &ft_null, AUDIO_MODULE_STATE_STOPPED, AUDIO_MODULE_STATE_RUNNING, 0);
+	test_start_stop(true, &ft_null, AUDIO_MODULE_STATE_CONFIGURED, AUDIO_MODULE_STATE_RUNNING,
+			0);
+}
+
+ZTEST(suite_audio_module_functional, test_start_fnct)
+{
+	test_start_stop(true, &ft_pop, AUDIO_MODULE_STATE_RUNNING, AUDIO_MODULE_STATE_RUNNING,
+			-EALREADY);
+	test_start_stop(true, &ft_pop, AUDIO_MODULE_STATE_STOPPED, AUDIO_MODULE_STATE_RUNNING, 0);
+	test_start_stop(true, &ft_pop, AUDIO_MODULE_STATE_CONFIGURED, AUDIO_MODULE_STATE_RUNNING,
+			0);
+}
+
+static void test_connections(bool connect, enum audio_module_type from_type,
+			     enum audio_module_type to_type, enum audio_module_state from_state,
+			     enum audio_module_state to_state)
+{
+	int ret;
+	int num_destinations;
+	char *test_inst_from_name = "TEST instance from";
+	char *test_inst_to_name = "TEST instance to";
+	struct audio_module_handle handle_from;
+	struct audio_module_handle handles_to[TEST_CONNECTIONS_NUM];
+
+	test_from_description.type = from_type;
+	test_to_description.type = to_type;
+
+	for (int k = 0; k < TEST_CONNECTIONS_NUM; k++) {
+		test_initialise_handle(&handles_to[k], &test_to_description, NULL, NULL);
+		handles_to[k].state = to_state;
+		snprintf(handles_to[k].name, CONFIG_AUDIO_MODULE_NAME_SIZE, "%s %d",
+			 test_inst_to_name, k);
+	}
+
+	test_initialise_handle(&handle_from, &test_from_description, NULL, NULL);
+	memcpy(&handle_from.name, test_inst_from_name, CONFIG_AUDIO_MODULE_NAME_SIZE);
+	handle_from.state = from_state;
+	sys_slist_init(&handle_from.handle_dest_list);
+	k_mutex_init(&handle_from.dest_mutex);
+
+	if (connect) {
+		for (int k = 0; k < TEST_CONNECTIONS_NUM; k++) {
+			ret = audio_module_connect(&handle_from, &handles_to[k], false);
+
+			zassert_equal(ret, 0,
+				      "Connect function did not return successfully: "
+				      "ret %d (%d)",
+				      ret, k);
+			zassert_equal(handle_from.dest_count, k + 1,
+				      "Destination count is not %d, %d", k + 1,
+				      handle_from.dest_count);
+		}
+
+		test_list(&handle_from, &handles_to[0], TEST_CONNECTIONS_NUM, false);
+	} else {
```

Comment:
I am looking at these. Is there a test that checks that connections cannot be made to a module which is already running?

Reply:
Or do we not impose this limitation?

Reply:
We do not impose this limitation. The only time (except for the usual) we would fail is if there is already a connection between the modules.